File: algorithms/sort/bubble_sort.go
package sort

func bubbleSort(nums []int) []int {
	end := len(nums)
	// if end == 1，表示只剩下一個元素可排序 = 不用排序
	for end > 1 {
		for i := 0; i < end-1; i++ {
			// 每次兩兩比較從頭看到尾，可確保最大的元素會被丟到後面
			if nums[i] > nums[i+1] {
				// 發現左邊比右邊大，swap them
				nums[i], nums[i+1] = nums[i+1], nums[i]
			}
		}
		// 故當我已經知道最大的元素被丟到後面，那麼下一次整理時，就可以不用看到最後一個
		end--
	}
	return nums
}

File: javascript/bubble_sort.js
const assert = require('assert');

function bubbleSort(arr) {
	if (!Array.isArray(arr) || arr.length <= 1) return arr;
    for (let end = arr.length -1; end > 0; end --) {
        for (let i =0; i < end; i++) {
            if (arr[i] > arr[i+1]) {
                // swap
                const tmp = arr[i];
                arr[i] =  arr[i+1];
                arr[i+1] = tmp;
            }
        }
    }
	return arr;
}


const cases = [
    { name: 'given',
        input: [5, 1, 3, 2, 4], expected: [1, 2, 3, 4, 5] },
	{ name: 'empty slice',
        input: [], expected: [] },
	{ name: 'single element',
        input: [1], expected: [1] },
	{ name: 'already sorted',
        input: [1, 2, 3, 4, 5], expected: [1, 2, 3, 4, 5] },
	{ name: 'reverse sorted',
        input: [5, 4, 3, 2, 1], expected: [1, 2, 3, 4, 5] },
	{ name: 'unsorted',
        input: [3, 1, 4, 1, 5, 9, 2, 6], expected: [1, 1, 2, 3, 4, 5, 6, 9] },
	{ name: 'duplicates',
        input: [2, 2, 1, 1], expected: [1, 1, 2, 2] },
];

for (const { name, input, expected } of cases) {
	const result = bubbleSort([...input]); // use copy to avoid mutating the test case input
	assert.deepStrictEqual(result, expected, `${name}: got ${result} want ${expected}`);
}

console.log('All bubbleSort tests passed');


